File: AppSets-iOS/AppSets/com/app/appsets/usecase/StartUseCase.swift
//
//  StartUseCase.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/15.
//

import Foundation

protocol SpotLightsState{
    
}

struct PinnedAppState : SpotLightsState {
    let items: [AppDefinition]
}

struct RecommendedItemState : SpotLightsState {
    let items: [ItemDefinition]
}

struct BingWallpaperState : SpotLightsState {
  let items: [MicrosoftBingWallpaper]
}

struct WordOfTheDayState : SpotLightsState {
    let items: [Any]
}

struct PopularSearchState : SpotLightsState {
    let icon:String
    let title:String
    let items:[Any]
}

struct ReversedSpaceState : SpotLightsState {
    
}

struct AudioPlayerState : SpotLightsState {
    
}

class StartUseCase : ObservableObject {
    
    private static let TAG = "StartUseCase"
    
    @Published var spotLightsState: [any SpotLightsState] = []
    
    func doInitData(_ context: Context) {
        loadSpotLight(context)
    }
    
    private func loadSpotLight(_ context: Context) {
        if(!spotLightsState.isEmpty){
            return
        }
        
        Task{
            if let spotLight = await AppSetsRepository().getSpotLight() {
                DispatchQueue.main.async {
                    let audioPlayer = AudioPlayerState()
                    self.spotLightsState.append(audioPlayer)
                    
                    let pinnedApps = PinnedAppState(items: [])
                    self.spotLightsState.append(pinnedApps)
                    
                    let recommendItems = RecommendedItemState(items: [])
                    self.spotLightsState.append(recommendItems)
                    
                    if let wallpaperList = spotLight.microsoftBingWallpaperList {
                        let bingWallpaper = BingWallpaperState(items: wallpaperList)
                        self.spotLightsState.append(bingWallpaper)
                    }
                    
                    var worldOfDayItems: [Any] = []
                    spotLight.wordOfTheDayList?.forEach({ wordOfDay in
                        worldOfDayItems.append(wordOfDay)
                    })
                    spotLight.todayInHistoryList?.forEach({ todayInHistory in
                        worldOfDayItems.append(todayInHistory)
                    })
                    let wordOfTheDay = WordOfTheDayState(items: worldOfDayItems)
                    
                    self.spotLightsState.append(wordOfTheDay)
                    
                    var popularSearchItems:[Any] = []
                    spotLight.popularSearches?.keywords?.forEach({ word in
                        popularSearchItems.append(word)
                    })
                    spotLight.baiduHotData?.hotsearch?.forEach({ hotsearch in
                        popularSearchItems.append(hotsearch)
                    })
                   
                    let popularSearch = PopularSearchState(icon: "", title: "热点", items: popularSearchItems)
                    
                    self.spotLightsState.append(popularSearch)
                    
                }
            }
        }
        
    }
    
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/usecase/StartUseCase.kt
package xcj.app.appsets.usecase

import android.content.Context
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.core.content.ContextCompat
import androidx.lifecycle.LiveData
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import xcj.app.appsets.db.room.repository.PinnedAppsRepository
import xcj.app.appsets.server.repository.AppSetsRepository
import xcj.app.appsets.ui.model.state.SpotLight
import xcj.app.compose_share.dynamic.ComposeLifecycleAware
import xcj.app.starter.android.AppDefinition
import xcj.app.starter.android.ItemDefinition
import xcj.app.starter.android.util.PackageUtil
import xcj.app.starter.android.util.PurpleLogger
import xcj.app.starter.server.requestRaw
import xcj.app.starter.test.LocalApplication

class StartUseCase(
    private val coroutineScope: CoroutineScope,
    private val pinnedAppsRepository: PinnedAppsRepository,
    private val appSetsRepository: AppSetsRepository,
) : ComposeLifecycleAware {

    companion object {
        private const val TAG = "StartUseCase"
    }

    val spotLightsState: MutableList<SpotLight> = mutableStateListOf()

    var allApps: MutableList<AppDefinition> = mutableStateListOf()

    val pinnedApp: MutableState<SpotLight.PinnedApp> =
        mutableStateOf(SpotLight.PinnedApp(emptyList()))

    val recommendItems: MutableState<SpotLight.RecommendedItem> =
        mutableStateOf(SpotLight.RecommendedItem(emptyList()))

    private val pinnedAppPackageNames: LiveData<List<String>> =
        pinnedAppsRepository.getAllPinnedAppsLiveData()

    init {
        pinnedAppPackageNames.observeForever {
            updatePinnedApps(allApps, it, pinnedApp)
        }
    }

    suspend fun doInitData(context: Context) {
        PurpleLogger.current.d(
            TAG,
            "doInitData, thread:${Thread.currentThread()}"
        )
        updateAllApps(allApps)
        updatePinnedApps(allApps, pinnedAppPackageNames.value, pinnedApp)
        updateRecommendItems(context, recommendItems)
        loadSpotLight(context)
    }

    private suspend fun updateAllApps(allApps: MutableList<AppDefinition>) {
        PackageUtil.getLauncherIntentAppDefinitionListFlow(LocalApplication.current).collect {
            allApps.addAll(it)
        }
    }

    private fun updateRecommendItems(
        context: Context,
        recommendItems: MutableState<SpotLight.RecommendedItem>,
    ) {
        val appDefinitionMutableList = allApps
        val forYouText = ContextCompat.getString(context, xcj.app.appsets.R.string.for_you)
        val mutableListOf = mutableListOf<ItemDefinition>().apply {
            val recommendItemsCount = (4..7).random()
            if (appDefinitionMutableList.size > recommendItemsCount) {
                while (size <= 5) {
                    val appDefinition = appDefinitionMutableList.random()
                    if (!contains(appDefinition)) {
                        appDefinition.description = forYouText
                        add(appDefinition)
                    }
                }
            } else if (appDefinitionMutableList.isNotEmpty()) {
                val elements = appDefinitionMutableList.onEach {
                    it.description = forYouText
                }
                addAll(elements)
            }
        }
        recommendItems.value = SpotLight.RecommendedItem(mutableListOf)
    }

    private fun updatePinnedApps(
        allApps: MutableList<AppDefinition>,
        pinnedAppPackageNames: List<String>?,
        pinnedApp: MutableState<SpotLight.PinnedApp>,
    ) {
        val pinnedAppDefinitionList = mutableListOf<AppDefinition>()
        for (app in allApps) {
            if (pinnedAppPackageNames?.find { pkName -> pkName == app.applicationInfo?.packageName } != null) {
                app.isPinned = true
                pinnedAppDefinitionList.add(app)
            }
        }
        pinnedApp.value = SpotLight.PinnedApp(pinnedAppDefinitionList)
    }

    fun addPinnedApp(pinnedAppPackageName: String?) {
        coroutineScope.launch(Dispatchers.IO) {
            pinnedAppsRepository.addPinnedApp(pinnedAppPackageName)
        }

    }

    fun unPinApp(appPackageName: String?) {
        if (appPackageName.isNullOrEmpty()) {
            return
        }
        coroutineScope.launch(Dispatchers.IO) {
            pinnedAppsRepository.unPinApp(appPackageName)
        }
    }


    fun onSnapShotPageStateClick(
        spotLight: SpotLight,
        context: Context,
        payload: Any?,
    ) {
        when (spotLight) {
            is SpotLight.PinnedApp -> {
                if (payload !is AppDefinition) {
                    return
                }
                val launchIntentForPackage =
                    payload.applicationInfo?.packageName?.let {
                        context.packageManager.getLaunchIntentForPackage(it)
                    }
                if (launchIntentForPackage != null) {
                    runCatching {
                        context.startActivity(launchIntentForPackage)
                    }
                }
            }

            is SpotLight.RecommendedItem -> {
                if (payload is ItemDefinition) {
                    if (payload !is AppDefinition) {
                        return
                    }
                    val launchIntentForPackage =
                        payload.applicationInfo?.packageName?.let {
                            context.packageManager.getLaunchIntentForPackage(it)
                        }
                    if (launchIntentForPackage != null) {
                        runCatching {
                            context.startActivity(launchIntentForPackage)
                        }
                    }
                }
            }

            else -> {}
        }
    }

    private fun loadSpotLight(context: Context) {
        if (spotLightsState.isNotEmpty()) {
            return
        }
        coroutineScope.launch {
            requestRaw(
                action = {
                    val response =
                        appSetsRepository.getSpotLight()
                    val spotLight = response.data ?: return@requestRaw

                    spotLight.microsoftBingWallpaperList?.let {
                        val bingWallpaper =
                            SpotLight.BingWallpaper(it)
                        spotLightsState.add(bingWallpaper)
                    }

                    val wordOfDayItems: MutableList<Any> = mutableListOf()
                    val wordOfDay = SpotLight.WordOfTheDay(wordOfDayItems)
                    spotLight.wordOfTheDayList?.let(wordOfDayItems::addAll)
                    spotLight.todayInHistoryList?.let(wordOfDayItems::addAll)
                    spotLightsState.add(wordOfDay)

                    val words: MutableList<Any> = mutableListOf()

                    val popularSearch = SpotLight.PopularSearch(
                        xcj.app.compose_share.R.drawable.ic_call_missed_outgoing_24,
                        ContextCompat.getString(context, xcj.app.appsets.R.string.hotspot),
                        words
                    )

                    spotLight.popularSearches?.keywords?.let {
                        words.addAll(it)
                    }

                    spotLight.baiduHotData?.hotsearch?.let {
                        words.addAll(it)
                    }

                    spotLightsState.add(popularSearch)

                }
            ).onFailure {
                PurpleLogger.current.d(TAG, "loadSpotLight, failed!")
            }
        }
    }

    override fun onComposeDispose(by: String?) {

    }
}